File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.js
function solution(genres, plays) {
    const ans = [];
    
    const playObj = {}; // genre : total
    const genreObj = {}; // genre: [ [idx, plays], ...]
    
    for(let i = 0; i < genres.length; i++){
        const genre = genres[i];
        const play = plays[i];
        
        if(!(genre in genreObj)){
            genreObj[genre] = [];
            playObj[genre] = 0;
        }
        
        genreObj[genre].push([i, play]);
        playObj[genre] += play;
    }
    
    //정렬
    const sortedGenres = Object.keys(playObj).sort((a,b) => playObj[b] - playObj[a]);
    for(const sortedGenre of sortedGenres){
        genreObj[sortedGenre].sort((a,b) => b[1] - a[1]);
        const top2 = genreObj[sortedGenre].slice(0,2).map((x) => x[0]);
        ans.push(...top2)
    }
    return ans;
}

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    ans = []
    total = {}
    songs = {}
    
    for i, (genre, play) in enumerate(zip(genres, plays)):
        total[genre] = total.get(genre, 0) + play
        songs.setdefault(genre, []).append((play, i))
    ordered_genres = sorted(total, key = lambda x : total[x], reverse=True)
    
    for genre in ordered_genres:
        
        songs[genre].sort(key = lambda x : (-x[0], x[1]))
        
        for play, index in songs[genre][:2]:
            ans.append(index)

    return ans